tests/utilities/test_auto_restart.py

```diff
@@ -1237,3 +1241,109 @@ def test_fault_tolerant_mode_enum():
     ):
         with mock.patch.dict(os.environ, {"PL_FAULT_TOLERANT_TRAINING": "3"}):
             _FaultTolerantMode.detect_current_mode()
+
+
+class StatefulRandomSampler(RandomSampler):
+
+    counter = 0
+
+    def state_dict(self):
+        self.counter += 1
+        return {"counter": self.counter}
+
+    def load_state_dict(self, state_dict):
+        self.counter = state_dict["counter"]
+
+
+class FailingStatefulRandomDataset(RandomDataset):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.counter = 0
+
+    def __getitem__(self, index):
+        self.counter += 1
+        return super().__getitem__(index)
+
+    def state_dict(self):
+        return {"counter": self.counter}
+
+    def load_state_dict(self, state_dict):
+        self.counter = state_dict["counter"]
+
+
+class StatefulRandomDataset(FailingStatefulRandomDataset):
+    def state_dict(self):
+        info = get_worker_info()
+        worker_id = info.id if info else 0
+        return {worker_id: {"counter": self.counter}}
+
+
+@pytest.mark.parametrize("num_workers", [0, 2])
```

Comment:
will you exclude num_workers=2 before merging?

Reply:
It takes 3 sec in total. I would prefer to include it as it covers the `_MultiProcessingDataLoaderIterStateful` functionalities.

Reply:
I know, but unfortunately 3 secs are an eternity considering we have thousands of tests. Recently we have hit our CI timeout of 40 minutes! cc @carmocca

Reply:
Fine for now, but let's re-explore the others auto restart tests. Some are taking 20 secs.

Reply:
Done !